### PR TITLE
Validating ISIN codes

### PR DESCRIPTION
**Pull request by raupachz**
An International Securities Identification Number (ISIN) uniquely identifies a security. Its structure is defined in ISO 6166. The isin uses country codes and Luhn algorithm for sanity checking.

### CODE REVIEW

**Comment by britter** [comment]
Nice! Since this is a bigger contribution I need to ask you to sign an [ICLA](https://www.apache.org/licenses/icla.txt) with the ASF

File: src/main/java/org/apache/commons/validator/routines/ISINValidator.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+import java.io.Serializable;
+import java.util.Locale;
+import org.apache.commons.validator.routines.checkdigit.LuhnCheckDigit;
+
+/**
+ * International Securities Identification Number (ISIN) uniquely identifies a
+ * bonds, commercial paper, stocks or warrant.
+ */
+public class ISINValidator implements Serializable {
```
Comment (britter) [inline src/main/java/org/apache/commons/validator/routines/ISINValidator.java]: There are several base classes in the routines package, which may be of help when implementing this. Have you looked at them? I don't know the validation logic of ISIN, but I could imagine that it has something to do with CheckDigits? Have a look at ISBNValidator and IBANValidator for examples.

File: src/main/java/org/apache/commons/validator/routines/ISINValidator.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+ */
+package org.apache.commons.validator.routines;
+
+import java.io.Serializable;
+import java.util.Locale;
+import org.apache.commons.validator.routines.checkdigit.LuhnCheckDigit;
+
+/**
+ * International Securities Identification Number (ISIN) uniquely identifies a
+ * bonds, commercial paper, stocks or warrant.
```
Comment (britter) [inline src/main/java/org/apache/commons/validator/routines/ISINValidator.java]: Maybe add a link to the ISO standard for ISIN?

**Comment by raupachz** [comment]
@britter I think I filed an CLA quite a while ago. Can you check with ASF? I checked the other methods and reuse the Luhn implementation. It is really a simple validation with everything that is already provided by Commons Validation. I will add a link to ISO 6166

File: src/main/java/org/apache/commons/validator/routines/ISINValidator.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+import java.io.Serializable;
+import java.util.Locale;
+import org.apache.commons.validator.routines.checkdigit.LuhnCheckDigit;
+
+/**
+ * International Securities Identification Number (ISIN) uniquely identifies a
+ * bonds, commercial paper, stocks or warrant. The structure is governed
+ * by ISO 6166.
+ */
```
Comment (sebbASF) [inline src/main/java/org/apache/commons/validator/routines/ISINValidator.java]: Is there a link to a public version of the document?

Comment (raupachz) [inline src/main/java/org/apache/commons/validator/routines/ISINValidator.java]: Must be this one: http://www.anna-web.org/standards/isin-iso-6166/

File: src/main/java/org/apache/commons/validator/routines/ISINValidator.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+        // check invariant
+        if (isin == null || isin.length() != ISIN_LENGTH || !isin.matches("^[A-Z0-9]{12}$")) {
+            return false;
+        }
+
+        // check country code or international code
+        String countryCode = isin.substring(0, 2);
+        if (!countryCode.equals("XS")) {
+            boolean foundCode = false;
```
Comment (sebbASF) [inline src/main/java/org/apache/commons/validator/routines/ISINValidator.java]: It's not clear why 'XS' is excluded - need to link to official docs.

**Comment by niallkp** [comment]
The code already exists in Commons Validator to do all this except for the ISO Country Code validation:

- We have an existing ISINCheckDigit routine (so no need to convert the characters and use the LuhnCheckDigit routine)

- We have the CodeValidator which combines regex validation, code length validation & check digit validation. So you can do something like the following:
`CodeValidator validator = new CodeValidator("[A-Z0-9]{12}", 12,ISINCheckDigit.ISIN_CHECK_DIGIT);`

Validation routines are more re-usable when then are split into small dedicated pieces. So I would suggest creating an "ISO Country Code Validator" and then combing that with the above CodeValidator & ISINCheckDigit to compose an "ISINValidator.

Lastly the Locale.getISOCountries() method does an "array copy" every  time it is called. So a new "ISO Country Code Validator" should call that method only once and cache the array to avoid the creation of the array every time.

**Comment by raupachz** [comment]
@niallkp Sorry, my bad. I totally missed the ISINCheckDigit routine. How, did I miss that one? In that case we probably don't need this pull request anymore. The parts are already there. Thank you for the detailed explanation. Keep up the good work!

**Comment by britter** [comment]
@raupachz okay, the let's just close this PR. Thank you!

**Pull request closed without merging**